pallets/moonbeam-lazy-migrations/src/foreign_asset.rs

```diff
@@ -0,0 +1,221 @@
+// Copyright 2024 Moonbeam foundation
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Functions for handling foreign asset migrations
+
+use super::*;
+use frame_support::sp_runtime::Saturating;
+use frame_support::traits::{fungibles::metadata::Inspect, ReservableCurrency};
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use sp_core::U256;
+
+#[derive(Debug, Encode, Decode, scale_info::TypeInfo, PartialEq, MaxEncodedLen)]
+pub enum ForeignAssetMigrationStatus {
+	/// No migration in progress
+	Idle,
```

Comment:
We could add intermediate steps here, like `MigrationStarted`, `MigratedBalances`, `MigratedApprovals` and `Completed`.

This way we can add extra safety checks on each function to ensure they are being called in the correct order.

And also, when the migration is finished, I would suggest using the `Completed` variant as it is more representative.

What do you think?

Reply:
Implementing this change will require significant time. Additionally, since the order of migration does not impact approvals and balances, either can be migrated first—balances before approvals or vice versa.